lib/rubocop/optimized_patterns.rb

```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+module RuboCop
+  # @api private
+  module OptimizedPatterns
+    # A wrapper around patterns array to perform optimized search.
```

Comment:
You can include a bit more of the reasoning for this class that you've mentioned in the PR description.